### PR TITLE
Adds aad support with unit tests.

### PR DESCRIPTION
**Pull request by njd5475**
This adds an optional parameter to encrypt/decrypt functions to support using the Additional Authenticated Data.

### CODE REVIEW

File: src/crypto.js [hunk 1]
```diff
@@ -76,7 +76,7 @@ export default function makeCryptoWith(opts) {
   _validateOpts(opts);
   const { encryptionKey } = opts;
 
-  function encrypt(input) {
+  function encrypt(input, aad) {
```
Comment (kobelb) [inline src/crypto.js]: Can we use the same serialization approach which we use for the `input` with the `aad`?

File: test/src/crypto.js [hunk 2]
```diff
@@ -153,4 +153,24 @@ describe('crypto', () => {
...
+    it('should not be decryptable without aad used in encryption', async () => {
+      const encrypted = await crypto.encrypt('Hello World!', 'some aad');
+      try {
+        const decrypted = await crypto.decrypt(encrypted);
+        throw new Error(`Decryption should fail but value was ${decrypted}`);
+      }catch(err) {
+        expect(err).to.be.an(Error);
+        expect(err.message).to.match(/Unsupported state or unable to authenticate data/);
+      }
+    });
```
Comment (tvernum) [inline test/src/crypto.js]: Sorry for the drive-by review, but I would feel more comfortable if there was an additional test with the AAD being a single byte different (not just missing). 
```
      const encrypted = await crypto.encrypt('some text', '123456789');
      try {
        const decrypted = await crypto.decrypt(encrypted, '123456780' );
```

Comment (kobelb) [reply test/src/crypto.js]: No apology needed! I considered making the same request, but didn’t initially because it seemed like it’d be testing node’s crypto library itself. However, it would also make me feel more comfortable.

Comment (tvernum) [reply test/src/crypto.js]: Crypto APIs are notorious for being confusing to use and behaving in unexpected ways. So, something like this is at risk of looking like the AAD value would be verified in the `setAAD` method, but actually it turns out that you need to call `verifyAAD` afterwards, or something like that.
So, I see this as testing that we're calling node's crypto correctly, rather than testing that the library works. If we wanted to do the latter, we'd need a bunch more tests, particularly with larger data sizes.

File: src/crypto.js (hunk 1 above)
Comment (njd5475) [reply src/crypto.js]: @kobelb The problem is guaranteeing ordering of keys and array elements with `JSON.stringify` we'd have to use something like `object-hash` or create one for use here. The aad buffer needs to be exactly the same bytes. Using a repeatable object hash can be done by the user of this package prior to passing in the AAD.

Comment (kobelb) [reply src/crypto.js]: If we're going to require it to be a string, can we add tests for this and ensure a user comprehensible Error is thrown?

File: test/src/crypto.js (hunk 2 above)
Comment (njd5475) [reply test/src/crypto.js]: @tvernum No need to apologize the more eyes the better. I added a test with the one byte difference. It passes so that is comforting. I don't see any methods or process in node-crypto to do verification of AAD.

File: src/crypto.js (hunk 1 above)
Comment (njd5475) [reply src/crypto.js]: @kobelb Ok I added a validation step for the AAD if it is defined.

File: src/crypto.js [hunk 3]
```diff
@@ -88,7 +98,7 @@ export default function makeCryptoWith(opts) {
       const [ serializedInput, iv, key ] = results;
       const cipher = crypto.createCipheriv(CIPHER_ALGORITHM, key, iv);
 
-      if(aad !== undefined) {
+      if(_validateAADType(aad)) {
```
Comment (kobelb) [inline src/crypto.js]: I find using `_valaidateAADType` to throw an error and return whether it's specified to be rather confusing. Can we change it to only do the validation and then check to see if we should specify it in a different step? Something like the following:

```
function _validateAAD(aad) {
  if (aad == null) {
     return;
  }

  if (typeof aad !== 'string') {
    throw new Error('AAD must be a string');
  }

  if (aad.length === 0) {
    throw new Error('AAD cannot be an empty string');
  }
}

_validateAAD(aad); // at the very top of encrypt and decrypt

if(aad != null) {
  cipher.setAAD(Buffer.from(aad, 'utf8'));
}
```

Comment (njd5475) [reply src/crypto.js]: @kobelb I don't have strong opinions either way. This way is fine by me.

File: src/crypto.js [hunk 4]
```diff
@@ -27,6 +27,20 @@ function _validateOpts({ encryptionKey }) {
   }
 }
 
+function _validateAADType(aad) {
```
Comment (kobelb) [inline src/crypto.js]:
```suggestion
function _validateAAD(aad) {
```
We're validating the value, type and length of the string.

File: src/crypto.js [hunk 5]
```diff
@@ -85,18 +99,23 @@ export default function makeCryptoWith(opts) {
       _generateKey(encryptionKey, salt)
     ])
     .then(results => {
+      _validateAADType(aad);
```
Comment (kobelb) [inline src/crypto.js]: Can we move this to the very beginning of the encrypt method? There's no reason for us to perform this work if the AAD is invalid.

Comment (njd5475) [reply src/crypto.js]: yeah, I was trying to think what happens if the error is thrown outside the promise but the recovery from the error is to change code and to call the function appropriately or pass the correct options. So might be better to throw an error on the line that calls encrypt instead of waiting till later.

**Comment by njd5475** [comment]
@kobelb and @tvernum Ok I hope I've addressed all the issues thus far. Any more things that might make this even better?

File: package.json [hunk 6]
```diff
@@ -28,7 +28,7 @@
     "babel-eslint": "6.1.2",
     "babel-plugin-add-module-exports": "0.2.1",
     "eslint": "3.4.0",
-    "eslint-plugin-mocha": "4.5.1",
+    "eslint-plugin-mocha": "4.4.0",
```
Comment (kobelb) [inline package.json]: Would you mind elaborating on why this change is required? Is this what caused us to add the `package-lock.json` above? I'm inclined to think this should be pulled out into a separate PR, but I might be missing some of the context.

File: test/src/crypto.js [hunk 7]
```diff
@@ -153,4 +153,136 @@ describe('crypto', () => {
       expect(await crypto.decrypt(encrypted)).to.be('I am a string');
     });
   });
+
+  describe('Using AAD', () => {
+    describe('type validation', () => {
+      describe('for encrypt', () => {
+        it('should throw an error if aad is empty string', async () => {
```
Comment (kobelb) [inline test/src/crypto.js]: `expect.js` has `throw`/`throwException`/`throwError`, can we not use those here, and elsewhere in these tests where decryption/encryption fails?

File: package.json (hunk 6 above)
Comment (njd5475) [reply package.json]: Ah yeah I wanted to pull it out into another PR as well, but it was required to make CI pass. But maybe I can pull this into another PR first.

File: test/src/crypto.js (hunk 7 above)
Comment (njd5475) [reply test/src/crypto.js]: Ah so it does, yeah I can refactor the tests to use those!

Comment (njd5475) [reply test/src/crypto.js]: I cannot seem to get expectjs to work with the promises. I had looked for a `rejected` option which it does not seem to support. Any suggestions because it would be nicer than the try...catch blocks

Comment (kobelb) [reply test/src/crypto.js]: > I cannot seem to get expectjs to work with the promises

😞 that's a bummer. We might be stuck doing our own solution, or migrating all tests to use `jest` or another maintained "assertion library". 

For the time being, how would you feel about introducing a `expectRejected` so we at least don't have to replicate this logic multiple times throughout this file? We could also potentially use the `finally` clause to make this a bit friendlier (this is a suggestion/nit, feel free to ignore it):


```
const expectRejected = async (promise, expectedError) => {
  let error = null;
  try {
    await promise;
  }  catch (err) {
    err = error;
  } finally {
    expect(er).not.to.be(null);
    expect(er).to.match(expectedError);
  }
}
```

Comment (njd5475) [reply test/src/crypto.js]: It'd be nice to switch to `jest` but I don't think it should be done in this PR. I don't particularly like adding custom expects functions unless they make the tests compatible with the future proposed syntax.

Comment (kobelb) [reply test/src/crypto.js]: > I don't particularly like adding custom expects functions unless they make the tests compatible with the future proposed syntax.

Do you have an approach in mind that would do so?

**Review by tvernum — APPROVED** [review]
The crypto & tests looks good to me.
I'll leave @kobelb to review the packaging etc.

File: test/src/crypto.js (hunk 7 above)
Comment (njd5475) [reply test/src/crypto.js]: I attempted to extend `expect.js` that did not go so well. I also took the time to update the package to use jest see #5 and can update these tests based on that PR and this one. I think that is the best direction to go in.

**Comment by kobelb** [comment]
@azasypkin do you mind giving this a quick review? I converted all of the mocha tests to jest after that PR landed in master.

File: src/crypto.test.js [hunk 8]
```diff
@@ -123,4 +123,76 @@ describe('crypto', () => {
...
+      describe('for decrypt', () => {
+        const encrypted = 'ImI9iJzw8dc3g47sM0fdJTa1+N+h9EI12kzFt1dzewuKx4E3rO6d6lqH+VSYNAf1H3m9HX/SFb0ZeJBx06NtERF2j9rY+1PtrOkJQfQCqnVC1PdlguZAMTG38vEfKBMgrJiK2+fqBFH80A==';
+
+        it('should throw an error if aad is empty string', async () => {
+          await expect(crypto.decrypt(encrypted, '')).rejects.toThrowError(/AAD cannot be an empty string/);
+        });
+
+        it('should be valid for string', async () => {
+          const decrypted = await crypto.decrypt(encrypted, 'should succeed');
+          expect(decrypted).not.toEqual(undefined);
```
Comment (azasypkin) [inline src/crypto.test.js]: __optional nit:__ `expect(encrypted).not.toBeUndefined();`

File: src/crypto.test.js [hunk 9]
```diff
@@ -123,4 +123,76 @@ describe('crypto', () => {
...
+  describe('Using AAD', () => {
+    describe('type validation', () => {
+      describe('for encrypt', () => {
+        it('should throw an error if aad is empty string', async () => {
+          await expect(crypto.encrypt('Hello World!', '')).rejects.toThrowError(/AAD cannot be an empty string/);
+        });
+
+        it('should be valid for string', async () => {
+          const encrypted = await crypto.encrypt('Hello World!', 'should succeed');
+          expect(encrypted).not.toEqual(undefined);
```
Comment (azasypkin) [inline src/crypto.test.js]: __optional nit:__ `expect(encrypted).not.toBeUndefined();`

File: src/crypto.test.js [hunk 10]
```diff
@@ -123,4 +123,76 @@ describe('crypto', () => {
...
+        it('should throw an error if AAD is an Object', async () => {
+          await expect(crypto.decrypt(encrypted, {})).rejects.toThrowError(/AAD must be a string/);
+        });
+
+        it('should throw an error if AAD is an function', async () => {
+          await expect(crypto.decrypt(encrypted, () => {})).rejects.toThrowError(/AAD must be a string/);
+        });
+      });
+    });
+
```
Comment (azasypkin) [inline src/crypto.test.js:179]: __optional nit:__ redundant new line.

File: src/crypto.test.js [hunk 11]
```diff
@@ -123,4 +123,76 @@ describe('crypto', () => {
...
+
+    it('should decryptable with same AAD', async () => {
```
Comment (azasypkin) [inline src/crypto.test.js]: __typo:__ `should decryptable` --> `should be decryptable`

**Review by azasypkin — APPROVED** [review]
LGTM! All nits are optional, feel free to ingore them.

Should we bump version in `package.json` before we publish updated package to NPM?

**Comment by kobelb** [comment]
> Should we bump version in package.json before we publish updated package to NPM? 

For sure, I was going to do so in a separate PR after merging this and the PR that switched the tests themselves to Jest, as doing so dropped support for Node 4 and Node 6.
